### PR TITLE
#25 Introduce a file object and proxy method calls to the disk to provide a more fluent API

### PR DESCRIPTION
**Pull request by grantcopley**

### CODE REVIEW

**Review by jclausen — CHANGES_REQUESTED** [review]
One small change and then this looks great on my end.

File: models/testing/AbstractDiskSpec.cfc [hunk 1]
```diff
@@ -1253,6 +1253,154 @@ component extends="coldbox.system.testing.BaseTestCase" {
 					} );
 				} );
 			} );
+
+			story( "We can work with a file object", function(){
+				beforeEach( function(){
+					var files = [ "some_file.txt", "another_file.txt" ];
```
[the remark this answers is not in the document]
Comment (grantcopley) [reply models/testing/AbstractDiskSpec.cfc:1259]: @jclausen Not following. What should I prefix it with - the name of the CFML engine?

**Review by jclausen — APPROVED** [review]

File: models/providers/RamProvider.cfc [hunk 2]
```diff
@@ -1271,42 +1271,6 @@ component accessors="true" extends="cbfs.models.AbstractDiskProvider" {
 		return contentsMap( argumentCollection = arguments );
 	}
 
-	/**************************************** STREAM METHODS ****************************************/
```
Comment (lmajano) [inline models/providers/RamProvider.cfc:1274]: I think this is heavy handed. The RAM disk can still produce streams and it was working and useful

File: models/AbstractDiskProvider.cfc [hunk 3]
```diff
@@ -252,4 +252,14 @@ component accessors="true" {
...
 	}
 
+	/**
+	 * Returns a File object for simplier API access
+	 *
+	 * @path The file path to check
+	 */
+	File function file( required path ){
+		arguments.disk = this;
+		return createObject( "component", "cbfs.models.File" ).init( argumentCollection = arguments );
```
Comment (lmajano) [inline models/AbstractDiskProvider.cfc:262]: Can you use the `new` notation please

```
new File( argumentCollection = arguments  )
```

File: models/File.cfc [hunk 4]
```diff
@@ -0,0 +1,333 @@
+component accessors="true" {
```
Comment (lmajano) [inline models/File.cfc:1]: Please add the header allf iles have for consistency, with purpose and copyright info

File: models/File.cfc [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+	property name="path";
+
+	property name="disk";
+
+	function init( required disk, required path ){
```
Comment (lmajano) [inline models/File.cfc:7]: Add cfc docs

File: models/File.cfc [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+	 */
+	function create(
+		required contents,
+		string visibility,
+		struct metadata   = {},
+		boolean overwrite = true,
+		string mode
+	){
+		arguments.path = getPath();
+		invoke( getDisk(), "create", arguments );
```
Comment (lmajano) [inline models/File.cfc:34]: Why didn't you just do `getDisk().create( argumentCollection = arguments )` much faster than `invoke`

File: models/File.cfc [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+	 *
+	 * @return File  - returns the copied file object
+	 *
+	 * @throws cbfs.FileNotFoundException - When the source doesn't exist
+	 * @throws cbfs.FileOverrideException - When the destination exists and no override has been provided
+	 */
+	function copy( required destination, boolean overwrite = true ){
+		arguments.source = getPath();
+		getDisk().copy( argumentCollection = arguments );
+		return createObject( "component", "cbfs.models.File" ).init(
```
Comment (lmajano) [inline models/File.cfc:116]: use the `new` approach please

File: models/File.cfc [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+	 *
+	 * @return File - the moved file object
+	 *
+	 * @throws cbfs.FileNotFoundException - When the source doesn't exist
+	 * @throws cbfs.FileOverrideException - When the destination exists and no override has been provided
+	 */
+	function move( required destination, boolean overwrite = true ){
+		arguments.source = getPath();
+		getDisk().move( argumentCollection = arguments );
+		return createObject( "component", "cbfs.models.File" ).init(
```
Comment (lmajano) [inline models/File.cfc:135]: same here

File: test-harness/tests/specs/providers/RamDiskSpec.cfc [hunk 9]
```diff
@@ -9,27 +9,6 @@ component extends="cbfs.models.testing.AbstractDiskSpec" {
 		super.run();
 
 		// Localized Suites
-
```
Comment (lmajano) [inline test-harness/tests/specs/providers/RamDiskSpec.cfc:12]: Please keep these

**Comment by lmajano** [comment]
Small changes and I can merge and release

**Pull request merged**